core/src/main/java/org/pentaho/platform/util/Emailer.java

```diff
@@ -261,89 +289,105 @@ public boolean send() {
-          final MimeMultipart oldMultipart = (MimeMultipart) msg.getContent();
-          final MimeMultipart newMultipart = new MimeMultipart( "related" );
-
-
-          for ( int i = 0; i < oldMultipart.getCount(); i++ ) {
-            BodyPart bodyPart = oldMultipart.getBodyPart( i );
-
-            final Object content = bodyPart.getContent();
-            //Main HTML body
-            if ( content instanceof String ) {
-              final String newContent = body + "<br/><br/>" + content;
-              final MimeBodyPart part = new MimeBodyPart();
-              part.setText( newContent, "UTF-8", "html" );
-              newMultipart.addBodyPart( part );
-            } else {
-              //CID attachments
-              newMultipart.addBodyPart( bodyPart );
-            }
-          }
-
+          setBodyForEmbeddedHtml( msg, body );
 
-          msg.setContent( newMultipart );
         }
       } else {
 
         // construct the message
         msg = new MimeMessage( session );
-        Multipart multipart = new MimeMultipart();
-
-        if ( body != null ) {
-          MimeBodyPart bodyMessagePart = new MimeBodyPart();
-          bodyMessagePart.setText( body, LocaleHelper.getSystemEncoding() );
-          multipart.addBodyPart( bodyMessagePart );
-        }
-
-        if ( attachment != null ) {
-          ByteArrayDataSource dataSource = new ByteArrayDataSource( attachment, attachmentMimeType );
-          // attach the file to the message
-          MimeBodyPart attachmentBodyPart = new MimeBodyPart();
-          attachmentBodyPart.setDataHandler( new DataHandler( dataSource ) );
-          attachmentBodyPart.setFileName( attachmentName );
-          multipart.addBodyPart( attachmentBodyPart );
-        }
+        Multipart multipart = setMultipart( body );
 
         // add the Multipart to the message
         msg.setContent( multipart );
       }
 
-      if ( from != null ) {
-        msg.setFrom( new InternetAddress( from, fromName ) );
-      } else {
-        // There should be no way to get here
-        logger.error( "Email.ERROR_0012_FROM_NOT_DEFINED" ); //$NON-NLS-1$
-      }
-
-      if ( ( to != null ) && ( to.trim().length() > 0 ) ) {
-        msg.setRecipients( Message.RecipientType.TO, InternetAddress.parse( to, false ) );
-      }
-      if ( ( cc != null ) && ( cc.trim().length() > 0 ) ) {
-        msg.setRecipients( Message.RecipientType.CC, InternetAddress.parse( cc, false ) );
-      }
-      if ( ( bcc != null ) && ( bcc.trim().length() > 0 ) ) {
-        msg.setRecipients( Message.RecipientType.BCC, InternetAddress.parse( bcc, false ) );
-      }
-
-      if ( subject != null ) {
-        msg.setSubject( subject, LocaleHelper.getSystemEncoding() );
-      }
-
-
+      setToFromSubject( from, fromName, to, cc, bcc, subject, msg );
       msg.setHeader( "X-Mailer", Emailer.MAILER ); //$NON-NLS-1$
       msg.setSentDate( new Date() );
 
-      Transport.send( msg );
+      if ( service != null ) {
```

Comment:
I don't under stand this scenario` service != null ` . Service will always be not null Emailer#setup would have thrown an exception at the very beginning.

The else statement seems unnecessary.

Reply:
@soagarwal1  i still don't understand this **else case** for:
```
if ( service != null ) {
  service.sendEmail( emailConfig, session, msg );
  } else {
   Transport.send( msg );
}
```

In current master for Emailer.java( https://github.com/pentaho/pentaho-platform/blob/b572b4a8cc604219147fdadab7e70d4f274e684f/core/src/main/java/org/pentaho/platform/util/Emailer.java ) There is no logic to check if service is null. 

We are almost guaranteed for it not to be null.

ActionUtil uses the default constructor for Emailer
https://github.com/pentaho/pentaho-platform/blob/master/core/src/main/java/org/pentaho/platform/util/ActionUtil.java#L415
`      Emailer emailer = new Emailer();`
We guaranteed not to get a null instance `PentahoSystem.get`, by our configuration in
https://github.com/pentaho/pentaho-platform/blob/master/assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/pentahoObjects.spring.xml#L45
`<bean id="IEmailService" class="org.pentaho.platform.plugin.services.email.EmailService" scope="session"/>`

If `IEmailService` bean was null, then Emailer#setup would throw a NullPointerException when populating the props variable, It accesses `service.getEmailConfig()`